ios/NotificationServiceExtension/NotificationService.swift

```diff
@@ -0,0 +1,239 @@
+    let avatar = fetchINImage(imageURL: notificationData.avatarURL, reportActionID: notificationData.reportActionID)
+    let sender = INPerson(personHandle: handle,
+                          nameComponents: nil,
+                          displayName: notificationData.userName,
+                          image: avatar,
+                          contactIdentifier: nil,
+                          customIdentifier: nil)
+    
+    // Configure the group/room name if there is one
+    var speakableGroupName: INSpeakableString? = nil
+    var recipients: [INPerson]? = nil
+    if (notificationData.roomName != nil) {
+      speakableGroupName = INSpeakableString(spokenPhrase: notificationData.roomName ?? "")
+      
+      // To add the group name subtitle there must be multiple recipients set. However, we do not have
+      // data on the participatns in the room/group chat so we just add a placeholder here. This shouldn't
+      // appear anywhere in the UI
+      let placeholderPerson = INPerson(personHandle: INPersonHandle(value: "placeholder", type: .unknown),
+                                       nameComponents: nil,
+                                       displayName: "placeholder",
+                                       image: nil,
+                                       contactIdentifier: nil,
+                                       customIdentifier: nil)
+      recipients = [sender, placeholderPerson]
+    }
+
+    // Because this communication is incoming, you can infer that the current user is
+    // a recipient. Don't include the current user when initializing the intent.
+    let intent = INSendMessageIntent(recipients: recipients,
+                                     outgoingMessageType: .outgoingMessageText,
+                                     content: notificationData.messageText,
+                                     speakableGroupName: speakableGroupName,
+                                     conversationIdentifier: String(notificationData.reportID),
+                                     serviceName: nil,
+                                     sender: sender,
+                                     attachments: nil)
+    
+    // When the group name is set, we force the avatar to just be the sender's avatar
+    intent.setImage(avatar, forParameterNamed: \.speakableGroupName)
+    
+    return intent
+  }
+  
+  override func serviceExtensionTimeWillExpire() {
+    // Called just before the extension will be terminated by the system.
+    // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
+    if let contentHandler = contentHandler, let bestAttemptContent =  bestAttemptContent {
+      contentHandler(bestAttemptContent)
+    }
+  }
+  
+  func fetchINImage(imageURL: String, reportActionID: String) -> INImage? {
+    guard let url = URL(string: imageURL) else {
+      return nil
+    }
+    
+    do {
+      let data = try Data(contentsOf: url)
```

Comment:
This is a sync call, do we need to be worried about it? I can see it's being used as an Extension.

Reply:
Yeah this is running in a separate process from the main application so this should be fine. There we're no UI hiccups while I tested notifications